crates/worker/src/partition_processor_manager.rs

```diff
@@ -96,12 +111,101 @@ enum AttachError {
+    task_id: TaskId,
     _created_at: MillisSinceEpoch,
     _key_range: RangeInclusive<PartitionKey>,
-    _control_tx: mpsc::Sender<PartitionProcessorControlCommand>,
+    planned_mode: RunMode,
+    handle: PartitionProcessorHandle,
     watch_rx: watch::Receiver<PartitionProcessorStatus>,
-    _task_id: TaskId,
+}
+
+impl ProcessorState {
+    fn new(
+        partition_id: PartitionId,
+        task_id: TaskId,
+        key_range: RangeInclusive<PartitionKey>,
+        handle: PartitionProcessorHandle,
+        watch_rx: watch::Receiver<PartitionProcessorStatus>,
+    ) -> Self {
+        Self {
+            partition_id,
+            task_id,
+            _created_at: MillisSinceEpoch::now(),
+            _key_range: key_range,
+            planned_mode: RunMode::Follower,
+            handle,
+            watch_rx,
+        }
+    }
+
+    async fn step_down(&mut self) -> Result<(), ShutdownError> {
+        if self.planned_mode != RunMode::Follower {
+            self.handle.step_down().await?;
+        }
+
+        self.planned_mode = RunMode::Follower;
+
+        Ok(())
+    }
+
+    async fn run_for_leader(
+        &mut self,
+        metadata_store_client: MetadataStoreClient,
+        node_id: GenerationalNodeId,
+    ) -> Result<(), Error> {
+        if self.planned_mode != RunMode::Leader {
+            let leader_epoch =
+                Self::obtain_next_epoch(metadata_store_client, self.partition_id, node_id).await?;
+            self.handle.run_for_leader(leader_epoch).await?;
+        }
+
+        self.planned_mode = RunMode::Leader;
+
+        Ok(())
+    }
+
+    async fn obtain_next_epoch(
+        metadata_store_client: MetadataStoreClient,
+        partition_id: PartitionId,
+        node_id: GenerationalNodeId,
+    ) -> Result<LeaderEpoch, ReadModifyWriteError> {
+        let epoch: EpochMetadata = metadata_store_client
+            .read_modify_write(partition_processor_epoch_key(partition_id), |epoch| {
+                let next_epoch = epoch
+                    .map(|epoch: EpochMetadata| epoch.claim_leadership(node_id, partition_id))
+                    .unwrap_or_else(|| EpochMetadata::new(node_id, partition_id));
+
+                Ok(next_epoch)
+            })
+            .await?;
+        Ok(epoch.epoch())
+    }
+}
+
+struct PartitionProcessorHandle {
+    control_tx: mpsc::Sender<PartitionProcessorControlCommand>,
+}
+
+impl PartitionProcessorHandle {
+    fn new(control_tx: mpsc::Sender<PartitionProcessorControlCommand>) -> Self {
+        Self { control_tx }
+    }
+
+    async fn step_down(&self) -> Result<(), ShutdownError> {
+        self.control_tx
+            .send(PartitionProcessorControlCommand::StepDown)
+            .await
+            .map_err(|_| ShutdownError)
+    }
+
+    async fn run_for_leader(&self, leader_epoch: LeaderEpoch) -> Result<(), ShutdownError> {
+        self.control_tx
+            .send(PartitionProcessorControlCommand::RunForLeader(leader_epoch))
+            .await
```

Comment:
I'm not sure if we should await in those cases. We don't want to block the control loop if a single processor is too busy, right?

Perhaps we can set a timeout at the call-site or switch to a try_send?

Reply:
Yes, this is indeed a problem. I am also not really happy with it right now. Switching it to `try_send` could be a first band-aid. I'll change it that way and open an issue to revisit the `PartitionProcessor` because there are also some other places where we await.